doc/authorization_policy_engine.md

```diff
@@ -0,0 +1,235 @@
+
+Here are the default set of rego policy and permissions/databindings. These are
+what is required to carry out the default authorization decisions as highlighted
+at the start of this document.
+
+#### Default policy.rego
+
+The default rego policy is located [here](/pkg/server/authpolicy/policy.rego).
+
+#### Default permissions.json (databindings)
+
+The default permissions.json is located [here](/pkg/server/authpolicy/permissions.json).
+
+The default permissions file contains a field called "apis".
+This field has a list of APIs that is current being configured with the rego
+policy.
+
+The fields of each object are as follows:
+
+| field             | Description | Example |
+| ----------------- | ----------- | ------- |
+| full_method       | The full method name of the API call | /spire.api.server.svid.v1.SVID/MintJWTSVID |
+| allow_any         | if true, sets result.allow to true | |
+| allow_local       | if true, sets result.allow_if_local to true | |
+| allow_admin       | if true, sets result.allow_if_admin to true | |
+| allow_downstream  | if true, sets result.allow_if_downstream to true | |
+| allow_agent       | if true, sets result.allow_if_agent to true | |
+
+# Extending the policy
+
+This section contains examples of how the authorization policy can be extended.
+
+## Example 1: Entry creation namespacing restrictions
+
+In this example, we want to ensure that entries created are namespaced, so we
+can create namespaces within the trust domain to determine the type of entries
+that can be created by each client. This would be a scenario of having two
+departments where one would not be able to create entries for the other.
+
+This can be defined by creating some additional objects in the data binding:
+
+```
+{
+    "entry_create_namespaces": [
+        {
+            "user": "spiffe://example.org/schedulers/finance",
+            "path_namespace": "/finance"
+        },
+        {
+            "user": "spiffe://example.org/schedulers/hr",
+            "path_namespace": "/hr"
+        }
+    ]
+}
+```
+
+The rego policy can then be updated to compare against the dataset of namespaces
+of users and path prefixes to compare against the entry create input request.
+```
+check_entry_create_namespace {
+    input.full_method == "/spire.api.server.entry.v1.Entry/BatchCreateEntry"
+
+    # caller has the registrar role
+    b = data.entry_create_namespaces[_]
+    b.user == input.caller
+
+    # spiffe id to be registered is in correct namespace
+    re_match(b.path_namespace, input.req.entries[_].spiffe_id.path)
+}
+
+check_entry_create_namespace {
+    input.full_method != "/spire.api.server.entry.v1.Entry/BatchCreateEntry"
+}
+```
+
+The rego policy can then be updated to check for this, an example of an allow
+clause would look like the following. Note that it is important to check to see
+how this fits in with the other parts of the rego policy.
+
+```
+# Any allow check
+allow = true {
+    check_entry_create_admin_flag
```

Comment:
Shouldn't it be `check_entry_create_namespace`?

I'm met with the following error if I use this line as it stands, which I surmise might arise from `check_entry_create_namespace` not being defined:

```bash
O[0000] Fatal run error                               error="unable to obtain authpolicy engine: 1 error occurred: spire.rego:97: rego_unsafe_var_error: var check_entry_create_admin_flag is unsafe"
ERRO[0000] Server crashed                                error="unable to obtain authpolicy engine: 1 error occurred: spire.rego:97: rego_unsafe_var_error: var check_entry_create_admin_flag is unsafe"
```

Reply:
yes - you are right, i copy-pasted and forgot to change 😓